proposals/src/ledger-replication-to-implement.md

```diff
@@ -37,3 +37,84 @@ transacation proves the validator incorrectly validated a fake storage proof.
 The replicator is rewarded and the validator's staking balance is slashed or
 frozen.
 
+### Storage proof contract logic
+
+Each replicator and validator will have their own storage account. The validator's
+account would be separate from their gossip id similiar to their vote account.
+These should be implemented as two programs one which handles the validator as the keysigner
+and one for the replicator. In that way when the programs reference other accounts, they
+can check the program id to ensure it is a validator or replicator account they are
+referencing.
+
+#### SubmitMiningProof
+```compile_fail
+SubmitMiningProof {
+    slot: u64,
+    sha_state: Hash,
+    signature: Signature,
+};
+keys = [replicator_keypair]
+```
+Replicators create these after mining their stored ledger data for a certain hash value.
+The slot is the end slot of the segment of ledger they are storing, the sha\_state
+the result of the replicator using the hash function to sample their encrypted ledger segment.
+The signature is the signature that was created when they signed a PoH value for the
+current storage epoch. The list of proofs from the current storage epoch should be saved
+in the account state, and then transfered to a list of proofs for the previous epcoh when
+the epoch passes. In a given storage epoch a given replicator should only submit proofs
+for one segment. The program should check that the slot and signature is valid and then store
+a record of this proof being submitted in the current storage epoch.
+
+#### ProofValidation
+```compile_fail
+ProofValidation {
+   proof_mask: Vec<ProofStatus>,
+}
+keys = [validator_keypair, replicator_keypair(s) (unsigned)]
```

Comment:
We also need to carry information about which proofs map to which replicator. I don't see how to do that here without some implicit ordering of account keys and proof_mask. Thats the only way to do it without adding more information in the proof_mask like we discussed earlier. That's the intended design, yes?

Reply:
Also some information about which segment this is for. Otherwise we'd have to pass in a proof_mask vec the size of the storage epoch where everything outside the current segment is `Skip`. Instead if we pass in the segment for which this validation was performed, the mask can be made tiny.

Reply:
The referenced replicator_keypair in `accounts[1]` will give them the set of storage proofs. Since a replicator keypair can only submit proofs for a single segment in a storage epoch time, then it specifies which segment it is for in that way. Could make it explicit, but I don't think it is strictly needed.